### PR TITLE
ci: use github actions cache to speedup bazel builds

### PR DESCRIPTION
**Pull request by malt3**
### Proposed change(s)

- ci: use github actions cache to speedup bazel build

### Additional info
- [run with empty cache](https://github.com/edgelesssys/constellation/actions/runs/4446625524) 11m15s
- [run with repo cache (but without remote cache)](https://github.com/edgelesssys/constellation/actions/runs/4446715094) 7m36s
- [run with repo cache and remote cache](https://github.com/edgelesssys/constellation/actions/runs/4468366094) 3m20s
- <strike>The github actions cache alone is a smaller improvement than I was hoping for. GitHub only allows uploading things to the cache if the cache key doesn't exist yet. We would want to merge the existing cache with new builds to get amazing speedups. See also: https://github.com/actions/cache/issues/342</strike>
- Combining caches and using a separate workflow to warm up the cache did the trick

### Checklist

- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [ ] Link to Milestone

### CODE REVIEW

File: .github/actions/setup_bazel/action.yml [hunk 1]
```diff
@@ -26,6 +26,17 @@ runs:
           exit 1
         fi
         echo "::endgroup::"
+    - name: Bazel repository cache (Linux)
```
Comment (katexochen) [inline .github/actions/setup_bazel/action.yml:29]:
```suggestion
        echo "::endgroup::"
        
    - name: Bazel repository cache (Linux)
```

File: .github/actions/setup_bazel/action.yml [hunk 2]
```diff
@@ -26,6 +26,17 @@ runs:
...
+      if: runner.os == 'Linux' && (inputs.useCache == 'true' || inputs.useCache == 'readonly')
+      with:
+        path: |
+          ${{ github.workspace }}/tools/pseudo-version
+          /home/runner/.cache/bazel
+          /home/runner/.cache/shared_bazel_repository_cache
+          /home/runner/.cache/shared_bazel_action_cache
+          /tmp/bazel-zig-cc
+        key: bazel
     - name: Configure Bazel
```
Comment (katexochen) [inline .github/actions/setup_bazel/action.yml:40]:
```suggestion
        key: bazel
        
    - name: Configure Bazel
```

File: .github/workflows/warm-bazel-cache.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            //upgrade-agent/cmd:upgrade_agent_linux_amd64
+
+      - name: clear
+        shell: bash
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          gh extension install actions/gh-actions-cache || true
+          gh actions-cache delete "${CACHE_KEY}" --confirm || true
+        continue-on-error: true
```
Comment (katexochen) [inline .github/workflows/warm-bazel-cache.yml:42]: This should be directly under name.

File: .github/actions/setup_bazel/action.yml [hunk 4]
```diff
@@ -26,6 +26,19 @@ runs:
           exit 1
         fi
         echo "::endgroup::"
+
+    - name: Bazel repository cache (Linux)
+      uses: actions/cache/restore@69d9d449aced6a2ede0bc19182fadc3a0a42d2b0 # v3.2.6
+      if: runner.os == 'Linux' && (inputs.useCache == 'true' || inputs.useCache == 'readonly')
```
Comment (katexochen) [inline .github/actions/setup_bazel/action.yml:32]:
```suggestion
      if: runner.os == 'Linux' && (inputs.useCache == 'true' || inputs.useCache == 'readonly')
      uses: actions/cache/restore@69d9d449aced6a2ede0bc19182fadc3a0a42d2b0 # v3.2.6
```

**Review by katexochen — APPROVED** [review]
Sorry, missed those two in the first rev.

**Review by daniel-weisse — APPROVED** [review]

**Pull request merged**